zetaclient/chains/ton/observer/outbound.go

```diff
@@ -0,0 +1,323 @@
+
+	return err
+}
+
+// return withdrawal and tx signer
+func extractWithdrawal(tx *toncontracts.Transaction) (toncontracts.Withdrawal, eth.Address, error) {
+	w, err := tx.Withdrawal()
+	if err != nil {
+		return toncontracts.Withdrawal{}, eth.Address{}, errors.Wrap(err, "not a withdrawal")
+	}
+
+	s, err := w.Signer()
+	if err != nil {
+		return toncontracts.Withdrawal{}, eth.Address{}, errors.Wrap(err, "unable to get signer")
+	}
+
+	return w, s, nil
+}
+
+// getOutboundByNonce returns outbound by nonce
+func (ob *Observer) getOutboundByNonce(nonce uint64) (outbound, bool) {
+	v, ok := ob.outbounds.Get(nonce)
+	if !ok {
+		return outbound{}, false
+	}
+
+	return v.(outbound), true
+}
+
+// setOutboundByNonce stores outbound by nonce
+func (ob *Observer) setOutboundByNonce(o outbound) {
+	ob.outbounds.Add(o.nonce, o)
+}
+
+func (ob *Observer) postVoteOutbound(
+	ctx context.Context,
+	cctx *cc.CrossChainTx,
+	w toncontracts.Withdrawal,
+	txHash string,
+	status chains.ReceiveStatus,
+) error {
+	// I. Gas
+	// TON implements a different tx fee model. Basically, each operation in our Gateway has a
+	// tx_fee(operation) which is based on hard-coded gas values per operation
+	// multiplied by the current gas fees on-chain. Each withdrawal tx takes gas directly
+	// from the Gateway i.e. gw pays tx fees for itself.
+	//
+	// - Gas price is stores in zetacore thanks to Observer.postGasPrice()
+	// - Gas limit should be hardcoded in TON ZRC-20
+	//
+	// II. Block height
+	// TON doesn't sequential block height because different txs might end up in different shard chains
+	// tlb.BlockID is essentially a workchain+shard+seqno tuple. We can't use it as a block height. Thus let's use 0.
+	// Note that for the sake of gas tracking, we use masterchain block height (not applicable here).
+	const (
+		outboundGasUsed     = 0
+		outboundGasPrice    = 0
+		outboundGasLimit    = 0
+		outboundBlockHeight = 0
+	)
+
+	var (
+		chainID       = ob.Chain().ChainId
+		nonce         = cctx.GetCurrentOutboundParam().TssNonce
+		signerAddress = ob.ZetacoreClient().GetKeys().GetOperatorAddress()
+		coinType      = cctx.InboundParams.CoinType
+	)
+
+	msg := cc.NewMsgVoteOutbound(
+		signerAddress.String(),
+		cctx.Index,
+		txHash,
+		outboundBlockHeight,
+		outboundGasUsed,
+		math.NewInt(outboundGasPrice),
+		outboundGasLimit,
+		w.Amount,
+		status,
+		chainID,
+		nonce,
+		coinType,
+	)
+
+	const gasLimit = gasconst.PostVoteOutboundGasLimit
+
+	var retryGasLimit uint64
+	if msg.Status == chains.ReceiveStatus_failed {
+		retryGasLimit = gasconst.PostVoteOutboundRevertGasLimit
+	}
+
+	log := ob.Logger().Outbound.With().
+		Uint64("outbound.nonce", nonce).
+		Str("outbound.outbound_tx_hash", txHash).
+		Logger()
+
+	zetaTxHash, ballot, err := ob.ZetacoreClient().PostVoteOutbound(ctx, gasLimit, retryGasLimit, msg)
+	if err != nil {
+		log.Error().Err(err).Msg("PostVoteOutbound: error posting vote")
+		return err
+	}
+
+	if zetaTxHash != "" {
```

Comment:
Shouldn't we log an error if `zetaTxHash` is empty?

Reply:
It's copied from other observers' implementation. zetaTxHash is `""` when this vote is already posted. We'll refine in orchestrator v2. 

https://github.com/zeta-chain/node/blob/ecb328e3d8b18afdeba151eb2e86c28c59616a40/zetaclient/zetacore/client_vote.go#L158